### ISSUE TITLE
[估值因子与动量叠加] | Value Momentum Overlay

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4497]
**标签**: quantitative-strategy
### 摘要
结合价值因子（市净率/市盈率）和动量因子构建多因子选股策略，通过因子正交化和IC动态加权提升预测能力。

### 算法描述

**核心原理**

价值因子和动量因子在学术和实务中均被证明具有显著的Alpha预测能力，但两者结合时存在信息重叠和因子拥挤问题。本策略通过正交化处理和动态权重优化实现因子互补。

**数学模型**

1. **价值因子**：
   $$PB = \frac{MarketCap}{BookValue}, \quad PE = \frac{Price}{Earnings}$$
   标准化：$V_{score} = \frac{PB - \mu_{PB}}{\sigma_{PB}}$

2. **动量因子**：
   $$Momentum_{12m} = \frac{P_t}{P_{t-12m}} - 1$$
   排除近1月：$Momentum_{11m} = \frac{P_t}{P_{t-1m}} - 1$（避免反转效应）

3. **因子正交化**：
   $$Momentum_{orth} = Momentum - \beta_{VM} \cdot Value + \epsilon$$
   回归残差作为纯净动量因子。

4. **IC（信息系数）计算**：
   $$IC_t = \text{Corr}(Factor_{rank}, FutureReturn_{t+1})$$
   滚动IC均值评估因子有效性。

5. **动态因子权重**：
   $$w_t = \frac{\overline{IC}_t / \sigma_{IC_t}}{\sum_j \overline{IC}_{j,t} / \sigma_{IC_{j,t}}}$$

6. **复合得分**：
   $$Score = \sum_{k=1}^{K} w_k \cdot F_k^{norm}$$

7. **行业市值中性化**：
   $$F^{neutral}_{i} = F_i - \sum_{j} \beta_j \cdot Industry_j - \gamma \cdot \ln(MarketCap_i)$$

**策略流程**：
1. 计算全市场股票价值因子和动量因子
2. 对因子进行行业市值中性化处理
3. 执行因子正交化去除冗余信息
4. 计算滚动IC确定因子权重
5. 构建复合得分选股池
6. 按风险平价分配持仓

### Python代码示例

```python
import numpy as np
import pandas as pd
from scipy import stats
from typing import Dict, List, Tuple, Optional
from sklearn.linear_model import LinearRegression

class FactorProcessor:
    """因子处理器"""
    
    def __init__(self):
        self.neutralize_industries = True
        self.winsorize_std = 3  # 3倍标准差截尾
        
    def winsorize(self, factor_values: pd.Series) -> pd.Series:
        """因子截尾处理"""
        mean = factor_values.mean()
        std = factor_values.std()
        
        lower = mean - self.winsorize_std * std
        upper = mean + self.winsorize_std * std
        
        return factor_values.clip(lower, upper)
    
    def standardize(self, factor_values: pd.Series) -> pd.Series:
        """Z-score标准化"""
        return (factor_values - factor_values.mean()) / (factor_values.std() + 1e-10)
    
    def industry_neutralize(self, factor_values: pd.Series,
                           market_caps: pd.Series,
                           industry_labels: pd.Series) -> pd.Series:
        """
        行业市值中性化
        F_neutral = F - beta_industry * Industry - gamma * ln(MarketCap)
        """
        # 对数市值
        log_mcap = np.log(market_caps + 1)
        
        # 行业虚拟变量回归
        industry_dummies = pd.get_dummies(industry_labels, prefix='ind')
        
        X = pd.concat([industry_dummies, log_mcap], axis=1)
        y = factor_values
        
        # OLS回归
        reg = LinearRegression()
        reg.fit(X, y)
        
        # 残差作为中性化因子
        neutralized = y - reg.predict(X)
        
        return pd.Series(neutralized.values, index=factor_values.index)
    
    def orthogonalize(self, factor_to_orth: pd.Series, 
                     factor_controls: List[pd.Series]) -> pd.Series:
        """
        因子正交化
        去除目标因子中受控制因子解释的部分
        """
        if not factor_controls:
            return factor_to_orth
            
        X = pd.concat(factor_controls, axis=1)
        y = factor_to_orth
        
        reg = LinearRegression()
        reg.fit(X.values, y.values)
        
        # 残差
        orthogonal = y - reg.predict(X.values)
        
        return pd.Series(orthogonal.values, index=factor_to_orth.index)


class ValueMomentumStrategy:
    """估值动量叠加策略"""
    
    def __init__(self, data_provider):
        self.data = data_provider
        self.processor = FactorProcessor()
        
        # 超参数
        self.lookback_momentum = 252  # 12个月
        self.exclusion_period = 21    # 排除近1月
        self.rolling_ic_window = 60   # IC滚动窗口
        self.rebalance_frequency = 21  # 月度调仓
        
        # 因子权重
        self.factor_weights = {'value': 0.5, 'momentum': 0.5}
        
        # IC历史
        self.ic_history = {'value': [], 'momentum': []}
        
    def calculate_value_factor(self, stocks: List[str], 
                              reference_date: datetime) -> pd.Series:
        """计算价值因子（市净率倒数+市盈率倒数）"""
        values = {}
        
        for stock in stocks:
            try:
                # 获取财务数据
                book_value = self.data.get_book_value(stock, reference_date)
                market_cap = self.data.get_market_cap(stock, reference_date)
                earnings = self.data.get_earnings(stock, reference_date)
                price = self.data.get_price(stock, reference_date)
                
                if book_value > 0 and market_cap > 0:
                    pb = market_cap / book_value
                    values[stock] = 1.0 / pb if pb > 0 else 0  # PB倒数
                else:
                    values[stock] = 0
                    
            except Exception as e:
                continue
        
        value_series = pd.Series(values)
        value_series = self.processor.winsorize(value_series)
        value_series = self.processor.standardize(value_series)
        
        return value_series
    
    def calculate_momentum_factor(self, stocks: List[str],
                                  reference_date: datetime) -> pd.Series:
        """计算动量因子（12个月，排除近1月）"""
        momentums = {}
        
        for stock in stocks:
            try:
                price_end = self.data.get_price(stock, reference_date)
                price_start = self.data.get_price(stock, reference_date - timedelta(days=self.lookback_momentum))
                price_1m_ago = self.data.get_price(stock, reference_date - timedelta(days=self.exclusion_period))
                
                if price_end and price_start and price_1m_ago:
                    # 12个月动量（排除近1月）
                    mom_12m = (price_end / price_start) - 1
                    mom_1m = (price_end / price_1m_ago) - 1
                    # 动量 = 12m - 1m（去除短期反转）
                    momentum = mom_12m - mom_1m
                    momentums[stock] = momentum
                else:
                    momentums[stock] = 0
                    
            except Exception:
                continue
        
        mom_series = pd.Series(momentums)
        mom_series = self.processor.winsorize(mom_series)
        mom_series = self.processor.standardize(mom_series)
        
        return mom_series
    
    def calculate_rolling_ic(self, factor: pd.Series, 
                            future_returns: pd.Series,
                            window: int = 60) -> float:
        """
        计算滚动信息系数
        IC = Corr(Factor_rank, FutureReturn_rank)
        """
        # 对齐数据
        common_idx = factor.index.intersection(future_returns.index)
        
        if len(common_idx) < window:
            return 0.0
        
        # 取最近window天的数据
        recent_factor = factor[common_idx].tail(window)
        recent_returns = future_returns[common_idx].tail(window)
        
        # Spearman相关系数
        ic, _ = stats.spearmanr(recent_factor, recent_returns)
        
        return ic if not np.isnan(ic) else 0.0
    
    def update_factor_weights(self, ic_history: Dict[str, List[float]]):
        """根据IC更新因子权重"""
        new_weights = {}
        ic_scores = {}
        
        for factor_name, ic_list in ic_history.items():
            if len(ic_list) >= 20:
                recent_ic = ic_list[-20:]
                mean_ic = np.mean(recent_ic)
                std_ic = np.std(recent_ic)
                
                if std_ic > 0:
                    ic_scores[factor_name] = mean_ic / std_ic
                else:
                    ic_scores[factor_name] = 0
            else:
                ic_scores[factor_name] = 0.5
        
        # 归一化权重
        total_score = sum(max(ic_scores[k], 0) for k in ic_scores)
        
        if total_score > 0:
            for k in ic_scores:
                new_weights[k] = max(ic_scores[k], 0) / total_score
        else:
            new_weights = {'value': 0.5, 'momentum': 0.5}
        
        self.factor_weights = new_weights
        
    def build_composite_score(self, value_factor: pd.Series,
                             momentum_factor: pd.Series,
                             market_caps: pd.Series,
                             industry_labels: pd.Series) -> pd.Series:
        """构建复合得分"""
        # 行业市值中性化
        value_neutral = self.processor.industry_neutralize(
            value_factor, market_caps, industry_labels
        )
        momentum_neutral = self.processor.industry_neutralize(
            momentum_factor, market_caps, industry_labels
        )
        
        # 动量对价值正交化
        momentum_orth = self.processor.orthogonalize(momentum_neutral, [value_neutral])
        
        # 标准化
        value_norm = self.processor.standardize(value_neutral)
        momentum_norm = self.processor.standardize(momentum_orth)
        
        # 复合得分
        composite = (self.factor_weights['value'] * value_norm + 
                    self.factor_weights['momentum'] * momentum_norm)
        
        return composite
    
    def select_stocks(self, composite_score: pd.Series,
                     top_pct: float = 0.1,
                     min_mcap: float = 1e9) -> List[str]:
        """按复合得分选股"""
        # 过滤小市值
        valid_stocks = composite_score[composite_score.index.isin(
            self.data.get_large_cap_stocks(min_mcap)
        )]
        
        # 取Top 10%
        n_select = max(int(len(valid_stocks) * top_pct), 10)
        
        top_stocks = valid_stocks.nlargest(n_select).index.tolist()
        
        return top_stocks
    
    def calculate_risk_parity_weights(self, stocks: List[str],
                                     covariance_matrix: pd.DataFrame) -> pd.Series:
        """风险平价加权"""
        n = len(stocks)
        
        if n == 0 or covariance_matrix.empty:
            return pd.Series()
        
        # 协方差矩阵
        cov = covariance_matrix.loc[stocks, stocks].values
        
        # 风险平价：每只股票的边际风险贡献相等
        # 简化：逆波动率加权
        volatilities = np.sqrt(np.diag(cov))
        inv_vol = 1.0 / (volatilities + 1e-10)
        weights = inv_vol / inv_vol.sum()
        
        return pd.Series(weights, index=stocks)
    
    def generate_portfolio(self, reference_date: datetime,
                          universe: List[str]) -> Dict:
        """生成目标组合"""
        # 计算因子
        value_factor = self.calculate_value_factor(universe, reference_date)
        momentum_factor = self.calculate_momentum_factor(universe, reference_date)
        
        # 获取市值和行业
        market_caps = self.data.get_market_caps(universe, reference_date)
        industry_labels = self.data.get_industry_labels(universe)
        
        # 构建复合得分
        composite = self.build_composite_score(value_factor, momentum_factor,
                                              market_caps, industry_labels)
        
        # 选股
        selected = self.select_stocks(composite)
        
        if not selected:
            return {'stocks': [], 'weights': [], 'factor_scores': {}}
        
        # 风险平价加权
        cov_matrix = self.data.get_return_covariance(selected, reference_date, 60)
        weights = self.calculate_risk_parity_weights(selected, cov_matrix)
        
        return {
            'stocks': selected,
            'weights': weights.to_dict(),
            'factor_scores': {
                'value': value_factor[selected].to_dict(),
                'momentum': momentum_factor[selected].to_dict()
            },
            'rebalance_date': reference_date
        }
    
    def backtest_strategy(self, start_date: datetime, end_date: datetime,
                         universe: List[str],
                         initial_capital: float = 10000000) -> Dict:
        """回测策略"""
        dates = pd.date_range(start_date, end_date, freq='D')
        capital = initial_capital
        positions = {}
        history = []
        
        for i, date in enumerate(dates):
            # 只在调仓日执行
            if i % self.rebalance_frequency != 0:
                continue
            
            # 生成目标组合
            target = self.generate_portfolio(date, universe)
            
            if not target['stocks']:
                continue
            
            # 再平衡
            for stock in list(positions.keys()):
                if stock not in target['stocks']:
                    # 卖出
                    price = self.data.get_price(stock, date)
                    if price:
                        capital += positions[stock] * price
                    del positions[stock]
            
            for stock, weight in target['weights'].items():
                target_value = capital * weight
                price = self.data.get_price(stock, date)
                
                if price:
                    positions[stock] = target_value / price
            
            history.append({
                'date': date,
                'capital': capital,
                'num_positions': len(positions),
                'stocks': list(positions.keys())
            })
            
            # 计算每日盈亏
            if i > 0:
                prev_date = dates[i-1]
                daily_return = 0
                
                for stock, shares in positions.items():
                    price_now = self.data.get_price(stock, date)
                    price_prev = self.data.get_price(stock, prev_date)
                    
                    if price_now and price_prev:
                        daily_return += shares * (price_now - price_prev)
                
                capital += daily_return
        
        return {
            'final_capital': capital,
            'total_return': (capital - initial_capital) / initial_capital,
            'num_rebalances': len(history),
            'history': history
        }


if __name__ == "__main__":
    # 测试
    np.random.seed(42)
    
    # 模拟数据
    stocks = [f'Stock_{i:03d}' for i in range(100)]
    dates = pd.date_range('2024-01-01', periods=500, freq='D')
    
    # 模拟因子值
    value_scores = pd.Series(np.random.randn(100), index=stocks)
    momentum_scores = pd.Series(np.random.randn(100), index=stocks)
    
    processor = FactorProcessor()
    value_processed = processor.standardize(processor.winsorize(value_scores))
    momentum_processed = processor.standardize(processor.winsorize(momentum_scores))
    
    # 正交化
    momentum_orth = processor.orthogonalize(momentum_processed, [value_processed])
    
    print(f"价值因子均值: {value_processed.mean():.4f}")
    print(f"动量因子（正交后）均值: {momentum_orth.mean():.4f}")
    print(f"相关系数: {np.corrcoef(value_processed, momentum_orth)[0,1]:.4f}")
